lib/zlib.toit

```diff
@@ -207,21 +208,53 @@ abstract class Coder_:
       this.uninit_
 
   /**
-  Returns a reader that can be used to read the compressed or decompressed data
-    output by the Encoder or Decoder.
+  A reader that can be used to read the compressed or decompressed data output
+    by the Encoder or Decoder.
   */
   reader/ZlibReader
 
-  read_ -> ByteArray?:
+  read_ --wait/bool -> ByteArray?:
     if closed_read_: return null
-    return zlib_read_ zlib_
+    result := zlib_read_ zlib_
+    while result and wait and result.size == 0:
+      state_ &= ~STATE_READY_TO_READ_
+      signal_.wait: state_ & STATE_READY_TO_READ_ != 0
+      result = zlib_read_ zlib_
+    state_ |= STATE_READY_TO_WRITE_
+    signal_.raise
+    return result
 
   close_read_ -> none:
+    state_ |= STATE_READY_TO_WRITE_
```

Comment:
Probably want to make this a critical_do section.
If a `task` is canceled, you still want to uninit_, and without a critical section that might not happen. iirc.